payara/Payara
# PAYARA-3832 Monitoring Console : Request Traces View

### Summary
This PR includes changes to:

* add a web API for request tracing data form the trace store (details view)
* add wild-card monitoring data queries (used to query request traces monitoring data for overview view)
* add JS for horizontal bar charts (as used by request trace overview)
* add JS for request trace detail view
* add basic preset for a request tracing page
* updates collection of request tracing data
* prunes some other data sources to the relevant data and unifies the series naming structure
* makes JS view components be based on models to allow easy reuse in different contexts while decoupling the code
* summarise multiple values for the same series during a collection cycle (both are counted, new tests added)
* numerous improvements in JS code organisation and interaction

### Request Tracing Views
The overview integrated in user customisable pages uses monitoring data that is collected.
The collection will _see_ each trace just once shortly after it had occurred. The series depends on the innermost span of each trace which usually is the operation of interest. Its operation name becomes the `@` in the series. For example:
```
ns:trace @:PUT_myWebMethod Duration
ns:trace @:GET_anotherWebMethod Duration
```
The namespace `ns:trace` and the metric `Duration` are known and fix parts in the data.
With the addition of wild-card queries a user can now ask for
```
ns:trace @:* Duration
```
This query will match any request trace series. In the overview charts each of the matching series is displayed as a bar. A positive side effect of this solution is that it can be applied to other data where wild-cards make sense to grab a collection of series and display them in the same graph. The wild-card can be applied for any of the tag values or the metric name:
```
ns:* Metric
ns:example *
ns:example @:* Metric
```
When clicking on the bar in the overview a detail view is displayed that shows data directly from the trace store. A specific web API and graph was added for this data.

### Documentation
I started documenting the data structures used in JS code. This is a _work in progress_ effort that will be continued in further PRs on the monitoring console. It is meant to be a easy to write documentation for developers to look at should they need to work on the JS parts. E.g. read it on github.

### Testing
The details below are given so that those not using the server on daily basis manage to test MC.

**Setting up the server and deploying MC**

1. Download Server from https://jenkins.payara.fish/job/PayaraQuickBuildAndTest/1448/artifact/appserver/distributions/payara/target/payara.zip
2. Extract the zip anywhere, switch into the `payara5/bin` folder
3. Start the server by running command `./asadmin start-domain` (see also https://blog.payara.fish/administering-payara-server-5-with-the-cli)
4. Open the Admin GUI at `http://localhost:4848/` in a browser
5. Download the `monitoring-console-webapp.war` attached to https://payara.atlassian.net/browse/PAYARA-3832
6. To deploy the war file in the Admin GUI klick _Applications_ in the left menu, click button _Deploy..._ in the listing showing. Select the downloaded file and press _Ok_
7. Open Monitoring Console at http://localhost:8080/monitoring-console-webapp/ in a browser

Now you should have MC running shown the _Core_ page.

**Testing Request Tracing**

1. Configure request tracing in the Admin GUI: Goto _Configurations => server-config_ in the left menu.
2. Choose _Request Tracing_ in the list of options
3. check _Enabled_
4. change _Target Count_ to `2` (Sampling Options)
5. change _Time Value_ to `20` and _Time Unit_ to `Seconds` (Sampling Options)
6. change _Threshold Value_ to `10` and _Threshold Unit_ to `Milliseconds` (Request Filters)
7. check _Reservoir Sampling_ (Trace Store)
8. press _Save_ button (top right)

Goto the `Request Tracing` page of MC by clicking on _Core_ title and selecting the page.
Open the side-panel by clicking on the cogs symbol in the top right icon bar.
After some time the chart should start to show a bar. Click on the bar to get to the details view.
Click bars in the detail view to see the trace span data in the side panel. 
To go back to the main view click the backwards arrow in the top left corner.
